### ISSUE TITLE
Can the 'new()' constraint on IPersistentState<TState> be dropped?

### ISSUE THREAD

**Issue by alexrp** [dotnet/orleans#6334]
It stands to reason that a constructor needs to be present for serializers to create instances of a type. However, this constraint goes a step further and requires the type and its constructor to be public, which leads to many types being unnecessarily exposed from an assembly. I am not personally aware of any serialization frameworks that require a type's constructor to be *public*, so perhaps this constraint could be dropped?

**Comment by ReubenBond** [comment]
Removing the constraint sounds fine to me, but note that the `new()` constraint does not require the type to be public, only the constructor.

For example:

``` C#
void Main()
{
	var x = new PersistentState<Foo>();
}

internal class Foo { public Foo() {} }

public class PersistentState<TState> where TState : new() { }
```

**Comment by alexrp** [comment]
Ah, my reading of the compiler error was wrong. That does actually mean my primary concern is dealt with. I suppose removing the constraint could still have value wrt exposing the constructor.

On a related note, there's bad news though: I was just searching around on this topic, and it turns out that `XmlSerializer` (which I know Orleans uses in the ADO.NET provider at least) [requires types to be public](https://stackoverflow.com/a/6156822)...

**Comment by ReubenBond** [comment]
I honestly didn't know that `XmlSerializer` was being used anywhere. At least it's only in that one storage provider.

**Comment by alexrp** [comment]
How do other storage providers do XML serialization? If they use something different, perhaps it would be worth unifying the strategies in the next major release.

**Comment by ReubenBond** [comment]
I'm not sure any other storage providers support XML. What we need is a more general purpose way to select serializers for storage, rather than requiring each provider to roll their own.

**Comment by srivathsah** [comment]
@ReubenBond  I think this still remains a problem as all the states in our domain aggregates are immutable records ie no public parameterless constructor (But serializable). So not sure how could this be dealt with. Should I just inject my own persistentState interface then ?

**Comment by veikkoeeva** [comment]
@alexrp, @ReubenBond ADO.NET persistence supports the Orleans binary, XML and JSON. Or put otherwise, it supports anything and reaches it to database so that binary formats are saved to binary type fields while XML and JSON to their respective fields (database may have improved storage and processing capabilities for these specific types).

I concur with what @ReubenBond noted about a general purpose way. Currently the ADO.NET provider does imlement a factory adapter that lets one to pick a (de)serializer based on call parameters and also inject arbitrary transoformations to objects during either persisting or loading (think of refactoring and other situations). It would be nice to rethink a bit of the Orleans system so that this could be done perhaps between gateway and siloes, between grains and between grains and storage. I'm specifically thinking scenarios where one might want to stream data to be persisted without transforming much.


As a smaller case, I would very much like to refactor the system so that it does not hardcode the providers as currently, but provide them via constructor injection so it would become possible to provide any (de)serializer one wants to use. The current system has harcoded `UseJson` etc. to mimick other providers and in the 1.5 era there wasn't an easy way to inject them (or provide as a function parameter).

**Comment by ReubenBond** [comment]
I opened a PR to remove the constraint. It's a compilation-breaking change for anyone who has a custom IPersistentState/IStorage/IGrainRuntime implementation, but the fix is simple (remove the constraint from the implementation)

**Comment by sergeybykov** [comment]
Resolved via #6351.
